src/main/java/com/google/api/gax/bundling/PushingBundler.java

```diff
@@ -0,0 +1,234 @@
+ */
+public final class PushingBundler<E> {
+
+  private final Runnable flushRunnable =
+      new Runnable() {
+        @Override
+        public void run() {
+          process(removeBundle());
+        }
+      };
+
+  private final ArrayList<BundlingThreshold<E>> thresholds;
+  private final ScheduledExecutorService executor;
+  private final Duration maxDelay;
+  private final ThresholdBundleReceiver<E> receiver;
+  private final BundlingFlowController<E> flowController;
+  private final BundleMerger<E> bundleMerger;
+
+  // Invariant:
+  // - lock gates all accesses to members below
+  // - currentOpenBundle and currentAlarmFuture are either both null or both non-null
+  private final ReentrantLock lock = new ReentrantLock();
+  private E currentOpenBundle;
+  private Future<?> currentAlarmFuture;
+
+  private PushingBundler(Builder<E> builder) {
+    this.thresholds = new ArrayList<>(builder.thresholds);
+    this.executor = Preconditions.checkNotNull(builder.executor);
+    this.maxDelay = Preconditions.checkNotNull(builder.maxDelay);
+    this.receiver = Preconditions.checkNotNull(builder.receiver);
+    this.flowController = Preconditions.checkNotNull(builder.flowController);
+    this.bundleMerger = Preconditions.checkNotNull(builder.bundleMerger);
+
+    resetThresholds();
+  }
+
+  public static class Builder<E> {
+    private Collection<BundlingThreshold<E>> thresholds;
+    private ScheduledExecutorService executor;
+    private Duration maxDelay;
+    private ThresholdBundleReceiver<E> receiver;
+    private BundlingFlowController<E> flowController;
+    private BundleMerger<E> bundleMerger;
+
+    private Builder() {}
+
+    public Builder<E> setThresholds(Collection<BundlingThreshold<E>> thresholds) {
+      this.thresholds = thresholds;
+      return this;
+    }
+
+    public Builder<E> setExecutor(ScheduledExecutorService executor) {
+      this.executor = executor;
+      return this;
+    }
+
+    public Builder<E> setMaxDelay(Duration maxDelay) {
+      this.maxDelay = maxDelay;
+      return this;
+    }
+
+    public Builder<E> setReceiver(ThresholdBundleReceiver<E> receiver) {
+      this.receiver = receiver;
+      return this;
+    }
+
+    public Builder<E> setFlowController(BundlingFlowController<E> flowController) {
+      this.flowController = flowController;
+      return this;
+    }
+
+    public Builder<E> setBundleMerger(BundleMerger<E> bundleMerger) {
+      this.bundleMerger = bundleMerger;
+      return this;
+    }
+
+    public PushingBundler<E> build() {
+      return new PushingBundler<>(this);
+    }
+  }
+
+  public static <E> Builder<E> newBuilder() {
+    return new Builder<>();
+  }
+
+  /**
+   * Immediately send contained elements to the {@code ThresholdBundleReceiver}.
+   */
+  public void flush() {
+    try {
+      // We wait on the executor so that any running processing
+      executorProcess(removeBundle()).get();
```

Comment:
Synced offline. Copying here so we don't loose it:

>imagine you send 2 messages. send(1); send(2); flush();
(1) can be sent in a different RPC than (2). So when flush() is called, the RPC for (1) might already be "in flight" (started but not completed)
The call to flush then makes an RPC for (2), wait for completion, and return
but it doesn't guarantee that the RPC for (1) has completed.